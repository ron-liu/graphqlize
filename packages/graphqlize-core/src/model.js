import {
	taskTry, applySpec, Box, prop, propEq, pipe, filter, evolve, map, path, taskOf, I, curry, concat, keys,
	pathSatisfies, isNil, not, both, pathEq, either, equals, propFn, when, ifElse, propSatisfies, K, any, find, all,
	notEquals, tap, __, contains, converge, assoc, merge, over, lensProp
} from "./util";
import {Kind} from 'graphql'
import {CustomScalars} from './types'
import Sequelize from 'sequelize'
import type {Field, SequelizeType} from "./types";
import type {Fn1} from "./basic-types";
import {TYPE_KIND} from "./constants";
import {printJson} from "./util/misc";

const getName = path(['name', 'value'])
const getValue = path(['value', 'value'])
const isKind = propEq('kind')
const isList = either(
	isKind(Kind.LIST_TYPE),
	both(
		isKind(Kind.NON_NULL_TYPE),
		propSatisfies(isKind(Kind.LIST_TYPE), 'type')
	)
)
const stripeList = pipe(
	ifElse(
		isList,
		pipe(
			when(isKind(Kind.NON_NULL_TYPE), prop('type')),
			prop('type')
		),
		I
	)
)

const getGraphqlizeType = propFn('type', pipe(
	stripeList,
	when(isKind(Kind.NON_NULL_TYPE), prop('type')),
	getName
))

const defaultScalarMappings: CustomScalars = {
	Int: {sequelizeType: Sequelize.INTEGER},
	Float: {sequelizeType: Sequelize.FLOAT},
	String: {sequelizeType: Sequelize.STRING},
	Boolean: {sequelizeType: Sequelize.BOOLEAN},
	ID: {sequelizeType: Sequelize.UUID}
}

export const getModels = (ast, option) => taskTry(() => {
	const customScalars = Box(option)
	.map(prop('customScalars'))
	.fold(merge(defaultScalarMappings))
	
	const allCustomScalarNames = keys(customScalars)
	
	const allEnumNames = Box(ast)
	.map(prop('definitions'))
	.map(filter(isKind('EnumTypeDefinition')))
	.fold(map(getName))
	
	const allValueObjectTypes = Box(ast)
	.map(prop('definitions'))
	.map(filter(isKind('ObjectTypeDefinition')))
	.map(filter(propFn('directives', pipe(map(getName), any(equals('valueObject'))))))
	.fold(map(getName))
	
	const allPersistenceObjectTypes = Box(ast)
	.map(prop('definitions'))
	.map(filter(isKind('ObjectTypeDefinition')))
	.map(filter(propFn('directives', pipe(
		map(getName),
		all(both(notEquals('valueObject'), notEquals('outSourcing')))
	))))
	.fold(map(getName))
	
	const fieldTypeToFieldKind = ast => typeName => {
		if(contains(typeName, allCustomScalarNames)) return 'scalar'
		else if(contains(typeName, allEnumNames)) return 'enum'
		else if(contains(typeName, allValueObjectTypes)) return 'valueObject'
		else if(contains(typeName, allPersistenceObjectTypes)) return 'relation'
		
		return undefined
	}
	
	const getSequelizeType : Fn1<Field, SequelizeType> = field => {
		if (field.fieldKind === 'relation') return undefined
		if (field.isList) return Sequelize.JSONB
		if (field.fieldKind === 'enum') return Sequelize.STRING
		if (field.fieldKind === 'valueObject') return Sequelize.JSONB
		return customScalars[field.graphqlizeType].sequelizeType
	}
	
	const getDirectives = map(applySpec({
		name: getName,
		arguments: pipe(
			prop('arguments'),
			map(applySpec({
				name: getName,
				value: getValue
			}))
		)
	}))
	
	const systemFields = [
		{
			name: 'createdAt',
			allowNullList: true,
			fieldKind: "scalar",
			graphqlizeType: 'DateTime',
			sequelizeType: Sequelize.DATE,
			primaryKey: false,
			isList: false,
			allowNull: true,
			isSystemField: true,
			isUnique: false,
		} ,
		{
			name: 'updatedAt',
			allowNullList: true,
			fieldKind: "scalar",
			graphqlizeType: 'DateTime',
			sequelizeType: Sequelize.DATE,
			primaryKey: false,
			isList: false,
			allowNull: true,
			isSystemField: true,
			isUnique: false,
		}
	]
	
	return Box(ast)
		.map(prop('definitions'))
		.map(filter(isKind('ObjectTypeDefinition')))
		.map(map(over(lensProp('fields'), map(over(lensProp('directives'), getDirectives)))))
		.map(map(applySpec({
			name: getName,
			interfaces: propFn('interfaces', map(getName)),
			directives: propFn('directives', getDirectives),
			modelKind: pipe(
				propFn('directives', map(getName)),
				find(either(equals('valueObject'), equals('outSourcing'))),
				when(isNil, K(TYPE_KIND.PERSISTENCE))
			),
			fields: propFn('fields', pipe(
				map(pipe(
					applySpec({
						name: getName,
						allowNullList: propFn('type', pipe(
							both(
								isKind(Kind.NON_NULL_TYPE),
								propSatisfies(isKind(Kind.LIST_TYPE), 'type')
							),
							not
						)),
						isUnique: either(
							propFn('directives', any(propEq('name', 'isUnique'))),
							pipe(getName, equals('id'))
						),
						isList: propFn('type', isList),
						allowNull: propFn('type', pipe(stripeList, isKind(Kind.NON_NULL_TYPE), not)),
						isSystemField: K(false),
						graphqlizeType: getGraphqlizeType,
						fieldKind: pipe(getGraphqlizeType, fieldTypeToFieldKind(ast)),
						primaryKey: pipe(getName, equals('id'))
					}),
					field => ({...field, sequelizeType: getSequelizeType(field)})
				)),
				concat(systemFields)
			))
		})))
		.fold(I)
})
