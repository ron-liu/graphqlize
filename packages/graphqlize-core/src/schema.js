import type {GraphqlizeOption, Schema, Model, Field, GenModelInputOption, Action} from './types'
import type {Fn1, Fn2, Fn3} from 'basic-types'
import {flatten, mergeWith, over, concat, lensProp, Box, prop, map, when, K, pipe, __, either, capitalize, join, tap} from "./util";
import {List} from 'immutable-ext'

const systemSchema : Schema = {
	types: [
		`enum OrderDirectionEnum { Asc Desc }`,
		`type _QueryMeta @noTable {count: Int!}`,
		`type File @noTable {
			id: ID!
			name: String!
			createdAt: DateTime
			updatedAt: DateTime
			size: Int!
			url: String!
			key: String!
		}`
	]
}

export const mergeSystemSchema : Fn1<GraphqlizeOption, GraphqlizeOption> =
	over(lensProp('schema'), mergeWith(concat, systemSchema))

const getFieldInput : Fn1< Action, Fn1<Field, string>>
= action => field => {
	const {fieldKind, graphqlizeType} = field
	if (fieldKind === 'valueObject') return Box(graphqlizeType).map(capitalize).fold(concat(__, 'Input'))
	if (fieldKind === 'relation') return Box(graphqlizeType).map(capitalize).map(concat(capitalize(action))).fold(concat(__, 'Input'))
	return graphqlizeType
}

const buildInput : Fn3<Action, Model, [string], string>
= (action, model, fields) => `input ${capitalize(action)}${model.name}Input {${fields.join(' ')}}`

const getInputField: Fn1<GenModelInputOption, Fn1<Field, string>>
= ({allowIdNull, allowFieldsOtherThanIdNull, action}) => field => {
	const {name, isList, allowNullList, allowNull, fieldKind, graphqlizeType} = field
	return Box(field)
		.map(getFieldInput(action))
		.map(when(
			either(
				K(name !== 'id' && !allowNull),
				K(name === 'id' && !allowIdNull)
			),
			concat(__, '!')
		))
		.map(when(K(isList), pipe(concat('['), concat(__, ']'))))
		.map(when(K(isList && !allowNullList && !allowFieldsOtherThanIdNull), concat(__, '!')))
		.fold(concat(`${name}:`))
}

const genModelInput : Fn2<GenModelInputOption, Model, string>
= option => model => Box(model)
	.map(prop('fields'))
	.map(map(getInputField(option)))
	.fold(x => buildInput(option.action, model, x))

const genCreateModelInput = genModelInput({allowIdNull: true, allowFieldsOtherThanIdNull: false, action: 'create'})
const genUpdateModelInput = genModelInput({allowIdNull: false, allowFieldsOtherThanIdNull: true, action: 'update'})
const genUpsertModelInput = genModelInput({allowIdNull: true, allowFieldsOtherThanIdNull: true, action: 'upsert'})
const genDeleteModelInput : Fn1<Model, string> = model => buildInput('delete', model, ['id:ID!'])

const genModelInputs : Fn1<Model, [string]>
= model => List.of(genCreateModelInput, genDeleteModelInput, genUpdateModelInput, genUpsertModelInput)
	.ap(List.of(model))
	.toArray()

export const genModelsInputs: Fn1<[Model], [string]>
= models => Box(models)
	.map(map(genModelInputs))
	.fold(flatten)