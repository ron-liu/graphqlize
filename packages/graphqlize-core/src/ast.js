import type {Fn1} from './basic-types'
import type {Schema, Ast} from './types'
import {when, pipe, path, isArrayLike, join} from './util'
import {parse} from 'graphql'
import Result from 'folktale/result'

// Schema -> Result Ast
export const schemaToAst: Fn1<Schema, Ast> = pipe(
	path([ 'types' ]),
	when(isArrayLike, join('\n')),
	x => Result.try(() => parse(x, {noLocation: true}))
)