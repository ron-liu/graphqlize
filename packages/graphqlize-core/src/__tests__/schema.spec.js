import {getModels} from '../model'
import {getAst} from '../ast'
import {genModelsInputs} from '../schema'

test('simple models inputs should ok', async () => {
	const types = [`
	type Person {
		id: ID
		name: String
	}
	`]
	const option = {schema: {types}, customScalar: []}
	const inputs = await getAst(option)
		.chain(x=>getModels(x, option))
		.map(genModelsInputs)
		.run()
		.promise()
	
	console.log(inputs)
	expect(inputs).toMatchSnapshot() // run: npm test -- -u, if code changed
})