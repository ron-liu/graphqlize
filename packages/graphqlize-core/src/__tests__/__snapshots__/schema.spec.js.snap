// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema to string should ok 1`] = `
"

type Post {name: String}

type Query {

}

type Mutation {

}

schema {   }"
`;

exports[`schema to string should ok 2`] = `
"

type Post {name: String}

type Query {

}

type Mutation {
createPost(name:String):Post
updatePost(id:ID, name:String):Post
}

schema {  mutation:Mutation }"
`;

exports[`schema to string should ok 3`] = `
"

type Post {name: String}

type Query {
posts(id:ID, name:String):[Post]
post(id:ID):Post
}

type Mutation {

}

schema { query:Query  }"
`;

exports[`schema to string should ok 4`] = `
"

type Post {name: String}

type Query {
posts(id:ID, name:String):[Post]
post(id:ID):Post
}

type Mutation {
createPost(name:String):Post
updatePost(id:ID, name:String):Post
}

schema { query:Query mutation:Mutation }"
`;

exports[`types to inputs inputs should ok 1`] = `
Array [
  "input CreatePersonInput {createdAt:DateTime, updatedAt:DateTime, id:ID, name:String, v_histories:[HistoryInput!], v_latestHistory:HistoryInput}",
  "input DeletePersonInput {id:ID!}",
  "input UpdatePersonInput {createdAt:DateTime, updatedAt:DateTime, id:ID!, name:String, v_histories:[HistoryInput!], v_latestHistory:HistoryInput}",
  "input UpsertPersonInput {createdAt:DateTime, updatedAt:DateTime, id:ID, name:String, v_histories:[HistoryInput!], v_latestHistory:HistoryInput}",
  "enum PersonFieldEnum {createdAt, updatedAt, id, name}",
  "input PersonOrderByInput { column: PersonFieldEnum, direction: OrderDirectionEnum }",
  "input PersonFilter {AND:[PersonFilter], OR:[PersonFilter], createdAt:DateTime, createdAt_gte:DateTime, createdAt_gt:DateTime, createdAt_lt:DateTime, createdAt_lte:DateTime, createdAt_in:[DateTime], createdAt_ne:DateTime, createdAt_between:[DateTime], createdAt_notBetween:[DateTime], createdAt_notIn:[DateTime], createdAt_like:String, createdAt_notLike:String, updatedAt:DateTime, updatedAt_gte:DateTime, updatedAt_gt:DateTime, updatedAt_lt:DateTime, updatedAt_lte:DateTime, updatedAt_in:[DateTime], updatedAt_ne:DateTime, updatedAt_between:[DateTime], updatedAt_notBetween:[DateTime], updatedAt_notIn:[DateTime], updatedAt_like:String, updatedAt_notLike:String, id:ID, name:String, name_gte:String, name_gt:String, name_lt:String, name_lte:String, name_in:[String], name_ne:String, name_between:[String], name_notBetween:[String], name_notIn:[String], name_like:String, name_notLike:String",
  "input HistoryInput {createdAt:DateTime, updatedAt:DateTime, id:ID, name:String}",
]
`;

exports[`types to inputs inputs should ok 2`] = `
Array [
  "input CreatePersonInput {createdAt:DateTime, updatedAt:DateTime, id:ID, name:String, histories:[CreateHistoryInput!], latestHistory:CreateHistoryInput}",
  "input DeletePersonInput {id:ID!}",
  "input UpdatePersonInput {createdAt:DateTime, updatedAt:DateTime, id:ID!, name:String, histories:[UpsertHistoryInput!], latestHistory:UpsertHistoryInput}",
  "input UpsertPersonInput {createdAt:DateTime, updatedAt:DateTime, id:ID, name:String, histories:[UpsertHistoryInput!], latestHistory:UpsertHistoryInput}",
  "enum PersonFieldEnum {createdAt, updatedAt, id, name}",
  "input PersonOrderByInput { column: PersonFieldEnum, direction: OrderDirectionEnum }",
  "input PersonFilter {AND:[PersonFilter], OR:[PersonFilter], createdAt:DateTime, createdAt_gte:DateTime, createdAt_gt:DateTime, createdAt_lt:DateTime, createdAt_lte:DateTime, createdAt_in:[DateTime], createdAt_ne:DateTime, createdAt_between:[DateTime], createdAt_notBetween:[DateTime], createdAt_notIn:[DateTime], createdAt_like:String, createdAt_notLike:String, updatedAt:DateTime, updatedAt_gte:DateTime, updatedAt_gt:DateTime, updatedAt_lt:DateTime, updatedAt_lte:DateTime, updatedAt_in:[DateTime], updatedAt_ne:DateTime, updatedAt_between:[DateTime], updatedAt_notBetween:[DateTime], updatedAt_notIn:[DateTime], updatedAt_like:String, updatedAt_notLike:String, id:ID, name:String, name_gte:String, name_gt:String, name_lt:String, name_lte:String, name_in:[String], name_ne:String, name_between:[String], name_notBetween:[String], name_notIn:[String], name_like:String, name_notLike:String, histories_some:HistoryFilter, histories_none:HistoryFilter, latestHistory:HistoryFilter, latestHistoryId:ID",
  "input CreateHistoryInput {createdAt:DateTime, updatedAt:DateTime, id:ID, name:String}",
  "input DeleteHistoryInput {id:ID!}",
  "input UpdateHistoryInput {createdAt:DateTime, updatedAt:DateTime, id:ID!, name:String}",
  "input UpsertHistoryInput {createdAt:DateTime, updatedAt:DateTime, id:ID, name:String}",
  "enum HistoryFieldEnum {createdAt, updatedAt, id, name}",
  "input HistoryOrderByInput { column: HistoryFieldEnum, direction: OrderDirectionEnum }",
  "input HistoryFilter {AND:[HistoryFilter], OR:[HistoryFilter], createdAt:DateTime, createdAt_gte:DateTime, createdAt_gt:DateTime, createdAt_lt:DateTime, createdAt_lte:DateTime, createdAt_in:[DateTime], createdAt_ne:DateTime, createdAt_between:[DateTime], createdAt_notBetween:[DateTime], createdAt_notIn:[DateTime], createdAt_like:String, createdAt_notLike:String, updatedAt:DateTime, updatedAt_gte:DateTime, updatedAt_gt:DateTime, updatedAt_lt:DateTime, updatedAt_lte:DateTime, updatedAt_in:[DateTime], updatedAt_ne:DateTime, updatedAt_between:[DateTime], updatedAt_notBetween:[DateTime], updatedAt_notIn:[DateTime], updatedAt_like:String, updatedAt_notLike:String, id:ID, name:String, name_gte:String, name_gt:String, name_lt:String, name_lte:String, name_in:[String], name_ne:String, name_between:[String], name_notBetween:[String], name_notIn:[String], name_like:String, name_notLike:String",
]
`;

exports[`types to inputs inputs should ok 3`] = `
Array [
  "input CreatePersonInput {createdAt:DateTime, updatedAt:DateTime, id:ID, name:String}",
  "input DeletePersonInput {id:ID!}",
  "input UpdatePersonInput {createdAt:DateTime, updatedAt:DateTime, id:ID!, name:String}",
  "input UpsertPersonInput {createdAt:DateTime, updatedAt:DateTime, id:ID, name:String}",
  "enum PersonFieldEnum {createdAt, updatedAt, id, name}",
  "input PersonOrderByInput { column: PersonFieldEnum, direction: OrderDirectionEnum }",
  "input PersonFilter {AND:[PersonFilter], OR:[PersonFilter], createdAt:DateTime, createdAt_gte:DateTime, createdAt_gt:DateTime, createdAt_lt:DateTime, createdAt_lte:DateTime, createdAt_in:[DateTime], createdAt_ne:DateTime, createdAt_between:[DateTime], createdAt_notBetween:[DateTime], createdAt_notIn:[DateTime], createdAt_like:String, createdAt_notLike:String, updatedAt:DateTime, updatedAt_gte:DateTime, updatedAt_gt:DateTime, updatedAt_lt:DateTime, updatedAt_lte:DateTime, updatedAt_in:[DateTime], updatedAt_ne:DateTime, updatedAt_between:[DateTime], updatedAt_notBetween:[DateTime], updatedAt_notIn:[DateTime], updatedAt_like:String, updatedAt_notLike:String, id:ID, name:String, name_gte:String, name_gt:String, name_lt:String, name_lte:String, name_in:[String], name_ne:String, name_between:[String], name_notBetween:[String], name_notIn:[String], name_like:String, name_notLike:String",
]
`;
